# **MÉTODO DE BAYES - RISCO DE CRÉDITO**

In [1]:
import pandas as pd

In [2]:
risco_credito_table = pd.read_csv("/content/risco_credito.csv") 

In [3]:
risco_credito_table

historia divida garantias     renda     risco
0           ruim   alta   nenhuma      0_15      alto
1   desconhecida   alta   nenhuma     15_35      alto
2   desconhecida  baixa   nenhuma     15_35  moderado
3   desconhecida  baixa   nenhuma  acima_35      alto
4   desconhecida  baixa   nenhuma  acima_35     baixo
5   desconhecida  baixa  adequada  acima_35     baixo
6           ruim  baixa   nenhuma      0_15      alto
7           ruim  baixa  adequada  acima_35  moderado
8            boa  baixa   nenhuma  acima_35     baixo
9            boa   alta  adequada  acima_35     baixo
10           boa   alta   nenhuma      0_15      alto
11           boa   alta   nenhuma     15_35  moderado
12           boa   alta   nenhuma  acima_35     baixo
13          ruim   alta   nenhuma     15_35      alto

## Separando uma tabela para os valores previsores e outro para as classes.

In [4]:
risco_credito_previsores = risco_credito_table.iloc[:,0:-1].values
# O primeiro e o ultimo parametro é exclusivo (ou seja, não entra).
# Cria uma tabela só com as colunas index, historia, divida, garantias e renda.
# O atributa .values retorna um tabela do tipo nparray
risco_credito_previsores[0]

array(['ruim', 'alta', 'nenhuma', '0_15'], dtype=object)

In [5]:
risco_credito_classes = risco_credito_table.iloc[:,-1].values
# Cria uma tabela com a coluna risco, somente.
# O atributa .values retorna um tabela do tipo nparray

## Aplicando os Label Encoder nos valores categóricos nominais.

In [6]:
from sklearn.preprocessing import LabelEncoder

Será necessário criar um objeto Ecoder para cada atributo previsor não numérico.

In [7]:
l_encoder_historia  = LabelEncoder()
l_encoder_divida    = LabelEncoder()
l_encoder_garantias = LabelEncoder()
l_encoder_renda     = LabelEncoder()

In [8]:
risco_credito_previsores[:,0] = l_encoder_historia.fit_transform(risco_credito_previsores[:,0])
risco_credito_previsores[:,1] = l_encoder_divida.fit_transform(risco_credito_previsores[:,1])
risco_credito_previsores[:,2] = l_encoder_garantias.fit_transform(risco_credito_previsores[:,2])
risco_credito_previsores[:,3] = l_encoder_renda.fit_transform(risco_credito_previsores[:,3])

**Primeira Coluna (historia):**

0 - Boa | 1 - Desconehcida | 2 - Ruim 

**Segunda Coluna (divida):**

0 - Alta | 1 - Baixa

**Terceira Coluna (garantias):**

0 - Adequada | 1 - Nenhuma

**Quarta Coluna (renda):**

0 - 0_15 | 1 - 15_35 | 2 - acima_35



In [9]:
risco_credito_previsores

array([[2, 0, 1, 0],
       [1, 0, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 2],
       [1, 1, 1, 2],
       [1, 1, 0, 2],
       [2, 1, 1, 0],
       [2, 1, 0, 2],
       [0, 1, 1, 2],
       [0, 0, 0, 2],
       [0, 0, 1, 0],
       [0, 0, 1, 1],
       [0, 0, 1, 2],
       [2, 0, 1, 1]], dtype=object)

## Salvando atributos previsores e classes num arquivo .pickle

In [10]:
import pickle

with open("risco_credito.pkl","wb") as f:
    pickle.dump([risco_credito_previsores,risco_credito_classes],f)

## Começando a usar o método...

In [11]:
from sklearn.naive_bayes import GaussianNB
# A realização do Algorítimo de Naive Bayes será feita pelo objeto GaussianNB

naive_baye_risco_credito = GaussianNB()
naive_baye_risco_credito.fit(risco_credito_previsores,risco_credito_classes)

GaussianNB()

Para descobrir o que cada valor do encoder significa é preciso comparar os valores com a tabela original.

# Testando Previsões

In [14]:
# historia - Boa(0), divida - Alta(0), garatias - Nenhuma(1), renda - maior_que_35(2)

previsao = naive_baye_risco_credito.predict([[0,0,1,2]])

previsao

array(['baixo'], dtype='<U8')

Ou seja, o risco é baixo para uma pessoa com:

historia - Boa(0), divida - Alta(0), garatias - Nenhuma(1), renda - maior_que_35(2)

### Mostrar as Classes

In [15]:
naive_baye_risco_credito.classes_

array(['alto', 'baixo', 'moderado'], dtype='<U8')

### Mostrar número de resgistros por classe

In [16]:
naive_baye_risco_credito.class_count_

array([6., 5., 3.])

Ou seja, 6 registros com risco alto, 5 registors com risco baixo e 3 registros com risco moderado

### Procentagem de cada classe no banco de dados

In [17]:
naive_baye_risco_credito.class_prior_

array([0.42857143, 0.35714286, 0.21428571])

Ou seja, 42% da base de dados possue risco alto, 35% possui risco baixo e 21% possui risco moderado